internal/namespace/typesystem.go

```diff
@@ -507,26 +497,80 @@ func (nts *TypeSystem) Validate(ctx context.Context) (*ValidatedNamespaceTypeSys
 					}
 				}
 			}
+
+			// Check the caveat, if any.
+			if allowedRelation.GetRequiredCaveat() != nil {
+				_, err := nts.resolver.LookupCaveat(ctx, allowedRelation.GetRequiredCaveat().CaveatName)
```

Comment:
> I _think_ see what's the need for _predefined_ as part of the resolver. Correct me if I'm wrong, but the idea is that, while validating the incoming type system as part of the service controllers, you want to make sure a caveat exists, and that can happen because either it was just added as as part of the `WriteSchema` call, or it already existed in the datastore.

In V1 it has to be part of the `WriteSchema` call. In v1alpha1, it does not, and so we have to support both cases (and other testing scenarios as well)

> However, isn't `WriteSchema` calls supposed to need to come with the full schema definition, and thus all caveat definitions will come as part of the request, regardless it existed before? If so, why the need for the fallback in the resolver to load a types from the datastore?

v1alpha1, which allows for partial updating

Reply:
> I _think_ see what's the need for _predefined_ as part of the resolver. Correct me if I'm wrong, but the idea is that, while validating the incoming type system as part of the service controllers, you want to make sure a caveat exists, and that can happen because either it was just added as as part of the `WriteSchema` call, or it already existed in the datastore.

In V1 it has to be part of the `WriteSchema` call. In v1alpha1, it does not, and so we have to support both cases (and other testing scenarios as well)

> However, isn't `WriteSchema` calls supposed to need to come with the full schema definition, and thus all caveat definitions will come as part of the request, regardless it existed before? If so, why the need for the fallback in the resolver to load a types from the datastore?

v1alpha1, which allows for partial updating

Reply:
we really need to deprecate `v1alpha1` 😄